### PR TITLE
Fix discrete callback evaluation modification

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes the handling of https://github.com/SciML/DiffEqBase.jl/pull/997

### CODE REVIEW

File: src/callbacks.jl [hunk 1]
```diff
@@ -603,8 +603,8 @@ end
         end
         integrator.u_modified = true
         callback.affect!(integrator)
-        if integrator.sol.prob isa DAEProblem || !(integrator.f.mass_matrix isa UniformScaling)
-            DiffEqBase.initialize_dae!(integrator)
+        if (integrator.sol.prob isa DAEProblem || !(integrator.f.mass_matrix isa UniformScaling)) && integrator.u_modified
```
Comment (ChrisRackauckas) [inline src/callbacks.jl:606]:
```suggestion
        if integrator.u_modified
```

**Comment by topolarity** [comment]
Does this also resolve https://github.com/SciML/OrdinaryDiffEq.jl/issues/2131?

**Comment by oscardssmith** [comment]
it doesn't, but it really should. I've gone through this in the debugger and it is calling the initialize_dae at the right time, it's just that the call isn't changing anything. I think somehow the call isn't updating Sundials with the new `p`...

File: src/callbacks.jl [hunk 2]
```diff
@@ -603,8 +603,8 @@ end
         end
         integrator.u_modified = true
         callback.affect!(integrator)
-        if integrator.sol.prob isa DAEProblem || !(integrator.f.mass_matrix isa UniformScaling)
-            DiffEqBase.initialize_dae!(integrator)
+        if integrator.u_modified
```
Comment (oscardssmith) [inline src/callbacks.jl:606]: isn't this check dead since `integrator.u_modified = true` is 2 lines above this?

Comment (ChrisRackauckas) [reply src/callbacks.jl:606]: No that's the point, you're supposed to set it in the callback if you want to override it.

Comment (oscardssmith) [reply src/callbacks.jl:606]: right. That makes sense. Is this ready to merge?

**Comment by oscardssmith** [comment]
closing in favor of https://github.com/SciML/DiffEqBase.jl/pull/1007

**Pull request closed without merging**